glm/ext/quaternion_exponential.inl

```diff
@@ -46,17 +48,34 @@ namespace glm
 		//To deal with non-unit quaternions
 		T magnitude = sqrt(x.x * x.x + x.y * x.y + x.z * x.z + x.w *x.w);
 
-		//Equivalent to raising a real number to a power
-		//Needed to prevent a division by 0 error later on
-		if(abs(x.w / magnitude) > static_cast<T>(1) - epsilon<T>() && abs(x.w / magnitude) < static_cast<T>(1) + epsilon<T>())
-			return qua<T, Q>(pow(x.w, y), 0, 0, 0);
+		if(abs(x.w / magnitude) > cos_one_over_two<T>())
+		{
+			//Scalar component is close to 1; using it to recover angle would lose precision
+			//Instead, we use the non-scalar components since sin() is accurate around 0
 
-		T Angle = acos(x.w / magnitude);
-		T NewAngle = Angle * y;
-		T Div = sin(NewAngle) / sin(Angle);
-		T Mag = pow(magnitude, y - static_cast<T>(1));
+			//Prevent a division by 0 error later on
+			T VectorMagnitude = x.x * x.x + x.y * x.y + x.z * x.z;
+			if (VectorMagnitude == static_cast<T>(0)) {
+				//Equivalent to raising a real number to a power
+				return qua<T, Q>(pow(x.w, y), 0, 0, 0);
+			}
 
-		return qua<T, Q>(cos(NewAngle) * magnitude * Mag, x.x * Div * Mag, x.y * Div * Mag, x.z * Div * Mag);
+			T SinAngle = sqrt(VectorMagnitude) / magnitude;
+			T Angle = asin(SinAngle);
+			T NewAngle = Angle * y;
+			T Div = sin(NewAngle) / SinAngle;
+			T Mag = pow(magnitude, y - static_cast<T>(1));
+			return qua<T, Q>(cos(NewAngle) * magnitude * Mag, x.x * Div * Mag, x.y * Div * Mag, x.z * Div * Mag);
+		}
+		else
```

Comment:
Redundant else due to return. Same below

Reply:
I feared that factoring the code would make it harder to read here. But you are right that it does duplicate quite a few formulas. I have pushed a new version that takes your remark into account.

Reply:
Are you sure this will compile like this? Angle is not defined below

Reply:
Of course! My bad, I should have run the tests before pushing. Fixed.

Reply:
Did you push the fix?

Reply:
Apparently I didn't :sweat: . Sorry for that.